doc/TAM/SAI-Proposal-TAM-stream-telemetry.md

```diff
@@ -0,0 +1,798 @@
+     * @flags CREATE_AND_SET
+     * @default 1
+     * @validonly SAI_TAM_TELEMETRY_ATTR_TAM_REPORTING_TYPE == SAI_TAM_REPORTING_TYPE_TIME_BASED
+     */
+    SAI_TAM_TELEMETRY_ATTR_REPORTING_INTERVAL,
+
+    /**
+     * @brief Tam telemetry reporting type
+     *
+     * @type sai_tam_reporting_type_t
+     * @flags CREATE_AND_SET
+     * @default SAI_TAM_REPORTING_TYPE_TIME_BASED
+     */
+    SAI_TAM_TELEMETRY_ATTR_TAM_REPORTING_TYPE,
+
+    /**
+     * @brief Tam telemetry reporting chunk size
+     *
+     * defines the size of reporting chunk, which means TAM will report to the collector every time
+     * if the report count reaches the chunk size.
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 1
+     * @validonly SAI_TAM_TELEMETRY_ATTR_TAM_REPORTING_TYPE == SAI_TAM_REPORTING_TYPE_COUNT_BASED
+     */
+    SAI_TAM_TELEMETRY_ATTR_REPORTING_CHUNK_SIZE,
+
+    /**
+     * @brief Tam telemetry cache size
+     *
+     * If the collector isn't ready to receive the report, this value indicates how many
+     * reports that can be cached. 0 means no cache which is the default behavior.
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_TAM_TELEMETRY_ATTR_CACHE_SIZE,
+
+} sai_tam_telemetry_attr_t;
+
+```
+
+``` c++
+
+sai_attr_list[0].id = SAI_TAM_TELEMETRY_ATTR_TAM_TYPE_LIST;
+sai_attr_list[0].value.objlist.count = 1;
+sai_attr_list[0].value.objlist.list[0] =  sai_tam_tel_type_obj;
+
+sai_attr_list[1].id = SAI_TAM_TELEMETRY_ATTR_COLLECTOR_LIST;
+sai_attr_list[1].value.objlist.count = 1;
+sai_attr_list[1].value.objlist.list[0] = sai_tam_collector_obj;
+
+sai_attr_list[2].id = SAI_TAM_TELEMETRY_ATTR_REPORTING_TYPE;
+sai_attr_list[2].value.s32 = SAI_TAM_REPORTING_TYPE_COUNT_BASED
+
+sai_attr_list[3].id = SAI_TAM_TELEMETRY_ATTR_REPORTING_CHUNK_SIZE;
+sai_attr_list[3].value.u32 = chunk_size; // STREAM_TELEMETRY_PROFILE:profile_name[chunk_size] on Config DB
+
+sai_attr_list[4].id = SAI_TAM_TELEMETRY_ATTR_CACHE_SIZE;
+sai_attr_list[4].value.u32 = cache_size; // STREAM_TELEMETRY_PROFILE:profile_name[cache_size] on Config DB
+
+attr_count = 5;
+
+sai_create_tam_telemetry_fn(&sai_tam_telemetry_obj, switch_id, attr_count, sai_attr_list);
+
+```
+
+#### Creating TAM counter subscription objects
+
+Based on the STREAM_TELEMETRY_GROUP on Config DB, to create corresponding counter subscription objects.
+
+``` c++
+
+// Create counter subscription list
+
+sai_attr_list[0].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_TEL_TYPE;
+sai_attr_list[0].value.oid = sai_tam_tel_type_obj;
+
+sai_attr_list[1].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_OBJECT_ID;
+sai_attr_list[1].value.oid = port_obj;
+
+sai_attr_list[2].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_STAT_ID;
+sai_attr_list[2].value.oid = SAI_PORT_STAT_IF_IN_OCTETS;
+
+sai_attr_list[3].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_LABEL;
+sai_attr_list[3].value.oid = index; // Element ID of the object in the IPFIX template

```

Comment:
we need to add explicit capability check here instead of leveraging the return value.

for capability query, we need to add a new API in saiobject.h:

```c
sai_status_t sai_query_stats_hft_capability(
        _In_ sai_object_id_t switch_id,
        _In_ sai_object_type_t object_type,
        _Inout_ sai_stat_hft_capability_list_t *stats_capability);
```

Then in saitypes.h, add the sai_stat_hft_capability_list_t following the sai_stat_capability_list_t, and we need to add the minimum polling interval in it.